File: data_structures/java/queue.java
class node<obj>{
	node<obj> next;
    node<obj> prev;
	obj data;
	node(node<obj> n,node<obj> p, obj a){
		next = n;
        prev = p;
		data = a;
	}
	void setnext(node<obj> n){
		next = n;
	}
	void setele(obj a){
		data = a;
	}
    void setprev(node<obj> p){
        prev = p;
    }
	obj getele(){
		return data;
	}
	node<obj> getnext(){
		return next;
	}
    node<obj> getprev(){
        return prev;
    }
}

class queue<obj>{//queue is nothing but a DLL with strict FIFO order
	/*note this is a generic class to use it, the declaration will be
	queue<Object(eg - Integer/String..)> Q = new queue<Object>();
	*/
	private node<obj> last;
    private node<obj> first;
	private int size;
	queue(){
		last = null;
        first = null;
		size=0;
	}
	void enqueue(obj c){
		node<obj> n = new node<obj>(last,null, c);
        if(last != null){ last.setprev(n); }
		last=n;
        if(size==0){
            first = n;
        }
		this.size++;
	}
    Object dequeue() throws NullPointerException{
        if(size==0){
            throw new NullPointerException("queue empty, cant remove");
        }
		node<obj> temp = first;
        first = first.getprev();
        if (first != null){ first.setnext(null); }
        this.size--;
        return temp;
    }
	Object getlast() throws NullPointerException{
		if(this.size == 0){
			throw new NullPointerException("queue empty");
		}
		return last.getele();
	}
    Object getfirst() throws NullPointerException{
		if(this.size == 0){
			throw new NullPointerException("queue empty");
		}
		return first.getele();
	}
	node<obj> getNext(node<obj> Current){
		return Current.next;
	}
    node<obj> getPrev(node<obj> Current){
        return Current.prev;
    }
	int size(){
		return this.size;
	}
    void traverse(){
        node<obj> temp = first;
        for (int i=0;i<size;i++){
            System.out.println(temp.getele());
            temp = temp.prev;
        }
    }
}

File: data_structures/python/queue.py
class node:
    def __init__(s, n, p, d):
        s.next = n
        s.prev = p
        s.data = d

class queue:
    def __init__(s):
        s.last = None
        s.first = None
        s.size = 0
    def enqueue(s, d):
        n = node(s.last, None, d)
        if (s.last!=None):
            s.last.prev = n
        s.last = n
        if ( s.size == 0):
            s.first = n
        s.size+=1
    def dequeue(s):
        if (s.size == 0):
            raise Exception("queue empty")
        temp = s.first
        s.first = s.first.prev
        if(s.first != None):
            s.first.next = None
        s.size-=1
        return temp
    def getlast(s):
        if (s.size==0):
            raise Exception("queue empty")
        return s.last.data
    def getfirst(s):
        if(s.size == 0):
            raise Exception("queue empty")
        return s.first.data
    def traverse(s):
        temp = s.first
        for i in range(s.size):
            print(temp.data, end=',')
            temp = temp.prev